contracts/pair_stable/src/utils.rs

```diff
@@ -0,0 +1,386 @@
+    pub return_amount: Uint128,
+    pub spread_amount: Uint128,
+}
+
+/// ## Description
+/// Returns the result of a swap in form of a [`SwapResult`] object. In case of error, returns [`ContractError`].
+/// ## Params
+/// * **storage** is an object of type [`Storage`].
+///
+/// * **env** is an object of type [`Env`].
+///
+/// * **config** is an object of type [`Config`].
+///
+/// * **offer_asset** is an object of type [`Asset`]. This is the asset that is being offered.
+///
+/// * **offer_pool** is an object of type [`Uint128`]. This is the total amount of offer assets in the pool.
+///
+/// * **ask_pool** is an object of type [`Uint128`]. This is the total amount of ask assets in the pool.
+///
+/// * **pools** is an array of [`Asset`] type items. These are the assets available in the pool.
+pub(crate) fn compute_swap(
+    storage: &dyn Storage,
+    env: &Env,
+    config: &Config,
+    offer_asset: &Asset,
+    offer_pool: &Asset,
+    ask_pool: &Asset,
+    pools: &[Asset],
+) -> Result<SwapResult, ContractError> {
+    let token_precision = get_precision(storage, &offer_asset.info)?;
+    let offer_amount = adjust_precision(
+        offer_asset.amount,
+        token_precision,
+        config.greatest_precision,
+    )?;
+
+    let new_ask_pool = calc_y(
+        &offer_asset.info,
+        &ask_pool.info,
+        offer_pool.amount.checked_add(offer_amount)?,
+        pools,
+        compute_current_amp(config, env)?,
+    )?;
+
+    let token_precision = get_precision(storage, &ask_pool.info)?;
+    let new_ask_pool = adjust_precision(new_ask_pool, config.greatest_precision, token_precision)?;
+    let return_amount = ask_pool.amount.checked_sub(new_ask_pool)?;
+
+    // We consider swap rate 1:1 in stable swap thus any difference is considered as spread.
+    let spread_amount = offer_asset.amount.saturating_sub(return_amount);
+
+    Ok(SwapResult {
+        return_amount,
+        spread_amount,
+    })
+}
+
+/// ## Description
+/// Accumulate token prices for the assets in the pool.
+/// ## Params
+/// * **deps** is an object of type [`Deps`].
+///
+/// * **env** is an object of type [`Env`].
+///
+/// * **config** is an object of type [`Config`].
+///
+/// * **pools** is an array of [`Asset`] type items. These are the assets available in the pool.
+pub fn accumulate_prices(
+    deps: Deps,
+    env: Env,
+    config: &mut Config,
+    pools: &[Asset],
+) -> Result<(), ContractError> {
+    let block_time = env.block.time.seconds();
+    if block_time <= config.block_time_last {
+        return Ok(());
+    }
+
+    let greater_precision = config.greatest_precision.max(TWAP_PRECISION);
+
+    let time_elapsed = Uint128::from(block_time - config.block_time_last);
+
+    let immut_config = config.clone();
+    for (from, to, value) in config.cumulative_prices.iter_mut() {
+        let offer_asset = from.with_balance(adjust_precision(
+            Uint128::from(1u8),
+            0u8,
+            greater_precision,
```

Comment:
Rather this one:
```pub struct AdjustedAsset {
    pub info: AssetInfo,
    pub amount: Uint128,
    decimals: u8,
}```

Reply:
The above proposal can't be used because a token precision is unknown at compile time. Another idea is to use ```Decimal256``` and its method ```from_atomics```

Reply:
- ```query_token_precision``` function can become part of AssetInfo as ```decimals``` method
- A trait can be implemented for Decimal256 which transforms it to Uint256 with specified precision
- ```adjust_precision``` function can be deleted